workflow/controller/scope.go

```diff
@@ -57,26 +60,71 @@ func (s *wfScope) resolveVar(v string) (interface{}, error) {
 	return nil, errors.Errorf(errors.CodeBadRequest, "Unable to resolve: {{%s}}", v)
 }
 
-func (s *wfScope) resolveParameter(v string) (string, error) {
-	val, err := s.resolveVar(v)
+func (s *wfScope) resolveParameter(p *wfv1.ValueFrom) (string, error) {
+	if p == nil {
+		return "", nil
+	}
+	if p.Parameter == "" && p.Expression == "" {
+		return "", nil
+	}
+	var val interface{}
+	var err error
+	if p.Expression != "" {
+		val, err = expr.Eval(p.Expression, s.getAllParamArtifact())
+	} else {
+		val, err = s.resolveVar(p.Parameter)
+	}
+
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("unable to resolve expression: %s", err)
```

Comment:
good work with error translation here:

tips:

* use %s or %q to include more debugging relating data, in this case the expression. As expression can be empty, or contain spaces, use %q to quote it.
* use %w to wrap the error rather than %v